tidb can not restart after create global binding

## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)
1. execute the sqls
```sql
create table t1(a int, b int, key (a)) partition by range (a) (partition p0 values less than (10), partition p1 values less than (20));
create global binding for select * from t1 using select * from t1 use index (a);
show global bindings;
```
2. restart tidb node
3. check the `tidb_stderr.log`
<!-- a step by step guide for reproducing the bug. -->

### 2. What did you expect to see? (Required)
1. tidb could restarted
### 3. What did you see instead (Required)
1.tidb can not restart
```log
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
    panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x313b7ce]

goroutine 1 [running]:
github.com/pingcap/tidb/executor.(*Compiler).Compile.func1()
    /home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/executor/compiler.go:72 +0x445
panic({0x4318e40, 0x6ec6870})
    /usr/local/go/src/runtime/panic.go:884 +0x212
github.com/pingcap/tidb/statistics/handle.(*Handle).GetPartitionStats(0xc0003fdb80?, 0x4f8da00?, 0x4f72898?, {0x0?, 0xc0012ef05c?, 0x2?})
    /home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/statistics/handle/handle.go:997 +0x2e
github.com/pingcap/tidb/statistics/handle.(*Handle).GetTableStats(...)
    /home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/statistics/handle/handle.go:992
github.com/pingcap/tidb/planner/core.(*PlanBuilder).buildDataSource(0xc000b384e0, {0x4fafbb0, 0xc0015c5920}, 0xc0012d5790, 0xc0015fa050)
    /home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/planner/core/logical_plan_builder.go:4456 +0x9ce
github.com/pingcap/tidb/planner/core.(*PlanBuilder).buildResultSetNode(0xc000b384e0, {0x4fafbb0?, 0xc0015c5920?}, {0x4fc96b0?, 0xc0015fa000?}, 0xf0?)
    /home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/planner/core/logical_plan_builder.go:380 +0x19d
github.com/pingcap/tidb/planner/core.(*PlanBuilder).buildJoin(0xc00166cf50?, {0x4fafbb0?, 0xc0015c5920?}, 0x2?)
    /home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/planner/core/logical_plan_builder.go:720 +0x71d
github.com/pingcap/tidb/planner/core.(*PlanBuilder).buildResultSetNode(0x0?, {0x4fafbb0?, 0xc0015c5920?}, {0x4fc8948?, 0xc0015ee870?}, 0x0?)
    /home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/planner/core/logical_plan_builder.go:367 +0x271
github.com/pingcap/tidb/planner/core.(*PlanBuilder).buildTableRefs(0xc000b384e0?, {0x4fafbb0?, 0xc0015c5920?}, 0x393d09e?)
    /home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/planner/core/logical_plan_builder.go:359 +0x85
github.com/pingcap/tidb/planner/core.(*PlanBuilder).buildSelect(0xc000b384e0, {0x4fafbb0, 0xc0015c5920}, 0xc0015ea360)
    /home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/planner/core/logical_plan_builder.go:3916 +0x6c7
github.com/pingcap/tidb/planner/core.(*PlanBuilder).Build(0xc000b384e0, {0x4fafbb0, 0xc0015c5920}, {0x4fc4080?, 0xc0015ea360?})
    /home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/planner/core/planbuilder.go:804 +0x745
github.com/pingcap/tidb/planner.buildLogicalPlan({0x4fafbb0, 0xc0015c5920}, {0x501e818?, 0xc0003fdb80}, {0x4fc4080, 0xc0015ea360}, 0xc000b384e0)
    /home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/planner/optimize.go:461 +0x12f
github.com/pingcap/tidb/planner.optimize({0x4fafbb0, 0xc0015c5920}, {0x501e818?, 0xc0003fdb80}, {0x4fc4080?, 0xc0015ea360?}, {0x4fe5b50, 0xc0015c59b0})
    /home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/planner/optimize.go:382 +0x473
github.com/pingcap/tidb/planner.Optimize({0x4fafbb0, 0xc0015c5920}, {0x501e818, 0xc0003fdb80}, {0x4fc4080, 0xc0015ea360}, {0x4fe5b50, 0xc0015c59b0})
    /home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/planner/optimize.go:245 +0xf11
github.com/pingcap/tidb/planner/core.(*PlanBuilder).buildExplain(0xc000b50000, {0x4fafbb0, 0xc0015c5920}, 0xc0015fa070)
    /home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/planner/core/planbuilder.go:4783 +0xd9
github.com/pingcap/tidb/planner/core.(*PlanBuilder).Build(0xc000b50000, {0x4fafbb0, 0xc0015c5920}, {0x4fc2c80?, 0xc0015fa070?})
    /home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/planner/core/planbuilder.go:779 +0x432
github.com/pingcap/tidb/planner.buildLogicalPlan({0x4fafbb0, 0xc0015c5920}, {0x501e818?, 0xc0003fdb80}, {0x4fc2c80, 0xc0015fa070}, 0xc000b50000)
    /home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/planner/optimize.go:461 +0x12f
github.com/pingcap/tidb/planner.optimize({0x4fafbb0, 0xc0015c5920}, {0x501e818?, 0xc0003fdb80}, {0x4fc2c80?, 0xc0015fa070?}, {0x4fe5b50, 0xc0015c59b0})
    /home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/planner/optimize.go:382 +0x473
github.com/pingcap/tidb/planner.Optimize({0x4fafbb0, 0xc0015c5920}, {0x501e818, 0xc0003fdb80}, {0x4fc2c80, 0xc0015fa070}, {0x4fe5b50, 0xc0015c59b0})
    /home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/planner/optimize.go:245 +0xf11
github.com/pingcap/tidb/executor.(*Compiler).Compile(0xc00166efc8, {0x4fafbb0, 0xc0015c5920}, {0x4fc8580, 0xc0015fa070?})
    /home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/executor/compiler.go:116 +0x6f8
github.com/pingcap/tidb/session.(*session).ExecuteStmt(0xc0003fdb80, {0x4fafbb0, 0xc0015c5920}, {0x4fc8580?, 0xc0015fa070})
    /home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/session/session.go:2171 +0x54e
github.com/pingcap/tidb/session.(*session).ExecuteInternal(0xc0003fdb80, {0x4fafbb0, 0xc0015c5920}, {0xc000869840, 0x3f}, {0x0, 0x0, 0x0})
    /home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/session/session.go:1674 +0x3f2
github.com/pingcap/tidb/bindinfo.getHintsForSQL({0x501e818, 0xc0003fdb80}, {0xc000ccb300, 0x29})
    /home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/bindinfo/handle.go:951 +0x177
github.com/pingcap/tidb/bindinfo.(*BindRecord).prepareHints(0xc0015d0bc0, {0x501e818, 0xc0003fdb80})
    /home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/bindinfo/bind_record.go:178 +0x1e7
github.com/pingcap/tidb/bindinfo.(*BindHandle).newBindRecord(0xc00053bf00, {0xc0015de140?, 0x1?})
    /home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/bindinfo/handle.go:723 +0xbcf
github.com/pingcap/tidb/bindinfo.(*BindHandle).Update(0xc00053bf00, 0x1)
    /home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/bindinfo/handle.go:173 +0x6c5
github.com/pingcap/tidb/domain.(*Domain).LoadBindInfoLoop(0xc000bca000, {0x501e818, 0xc0003fdb80}, {0x501e818, 0xc000b26000})
    /home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/domain/domain.go:1444 +0xe5
github.com/pingcap/tidb/session.BootstrapSession({0x4fd95f0, 0xc000ea44b0})
    /home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/session/session.go:3301 +0x648
main.createStoreAndDomain()
    /home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/tidb-server/main.go:314 +0x1cb
main.main()
    /home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/tidb-server/main.go:214 +0x2ca
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
    panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x313b7ce]
```


### 4. What is your TiDB version? (Required)
```sql
mysql> select tidb_version();
+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| tidb_version()                                                                                                                                                                                                                                                                               |
+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Release Version: v6.5.0
Edition: Community
Git Commit Hash: 706c3fa3c526cdba5b3e9f066b1a568fb96c56e3
Git Branch: heads/refs/tags/v6.5.0
UTC Build Time: 2022-12-27 03:50:44
GoVersion: go1.19.3
Race Enabled: false
TiKV Min Version: 6.2.0-alpha
Check Table Before Drop: false
Store: tikv |
+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
1 row in set (0.00 sec)
```
<!-- Paste the output of SELECT tidb_version() -->

--- a/executor/analyzetest/analyze_test.go
+++ b/executor/analyzetest/analyze_test.go
@@ -2833,7 +2833,7 @@ PARTITION BY RANGE ( a ) (
 	tk.MustQuery("select * from t where a > 1 and b > 1 and c > 1 and d > 1")
 	require.NoError(t, h.LoadNeededHistograms())
 	tbl := h.GetTableStats(tableInfo)
-	require.Equal(t, 4, len(tbl.Columns))
+	require.Equal(t, 0, len(tbl.Columns))
 
 	// ignore both p0's 3 buckets, persisted-partition-options' 1 bucket, just use table-level 2 buckets
 	tk.MustExec("analyze table t partition p0")

--- a/planner/core/collect_column_stats_usage_test.go
+++ b/planner/core/collect_column_stats_usage_test.go
@@ -331,6 +331,11 @@ func TestCollectHistNeededColumns(t *testing.T) {
 		if len(tt.pruneMode) > 0 {
 			s.ctx.GetSessionVars().PartitionPruneMode.Store(tt.pruneMode)
 		}
+		if s.ctx.GetSessionVars().IsDynamicPartitionPruneEnabled() {
+			s.ctx.GetSessionVars().StmtCtx.UseDynamicPruneMode = true
+		} else {
+			s.ctx.GetSessionVars().StmtCtx.UseDynamicPruneMode = false
+		}
 		stmt, err := s.p.ParseOneStmt(tt.sql, "", "")
 		require.NoError(t, err, comment)
 		err = Preprocess(context.Background(), s.ctx, stmt, WithPreprocessorReturn(&PreprocessorReturn{InfoSchema: s.is}))

--- a/planner/core/integration_partition_test.go
+++ b/planner/core/integration_partition_test.go
@@ -1619,3 +1619,39 @@ func TestPartitionRangeColumnPruning(t *testing.T) {
 	tk.MustQuery(`select * from t1 where a = 'a' AND c = 'd'`).Check(testkit.Rows("a <nil> d"))
 	tk.MustExec(`drop table t1`)
 }
+
+func TestPartitionProcessorWithUninitializedTable(t *testing.T) {
+	store := testkit.CreateMockStore(t)
+	tk := testkit.NewTestKit(t, store)
+	tk.MustExec("use test")
+	tk.MustExec(" create table q1(a int, b int, key (a)) partition by range (a) (partition p0 values less than (10), partition p1 values less than (20));")
+	tk.MustExec(" create table q2(a int, b int, key (a)) partition by range (a) (partition p0 values less than (10), partition p1 values less than (20));")
+
+	rows := [][]interface{}{
+		{"HashJoin"},
+		{"├─PartitionUnion(Build)"},
+		{"│ ├─TableReader"},
+		{"│ │ └─TableFullScan"},
+		{"│ └─TableReader"},
+		{"│   └─TableFullScan"},
+		{"└─PartitionUnion(Probe)"},
+		{"  ├─TableReader"},
+		{"  │ └─TableFullScan"},
+		{"  └─TableReader"},
+		{"    └─TableFullScan"},
+	}
+	tk.MustQuery("explain format=brief select * from q1,q2").CheckAt([]int{0}, rows)
+
+	tk.MustExec("analyze table q1")
+	tk.MustQuery("explain format=brief select * from q1,q2").CheckAt([]int{0}, rows)
+
+	tk.MustExec("analyze table q2")
+	rows = [][]interface{}{
+		{"HashJoin"},
+		{"├─TableReader(Build)"},
+		{"│ └─TableFullScan"},
+		{"└─TableReader(Probe)"},
+		{"  └─TableFullScan"},
+	}
+	tk.MustQuery("explain format=brief select * from q1,q2").CheckAt([]int{0}, rows)
+}

--- a/planner/core/logical_plan_builder.go
+++ b/planner/core/logical_plan_builder.go
@@ -4497,6 +4497,8 @@ func (b *PlanBuilder) buildDataSource(ctx context.Context, tn *ast.TableName, as
 						fmt.Errorf("disable dynamic pruning due to %s has no global stats", tableInfo.Name.String()))
 				}
 			}
+		} else {
+			b.optFlag = b.optFlag | flagPartitionProcessor
 		}
 		pt := tbl.(table.PartitionedTable)
 		// check partition by name.

--- a/planner/core/logical_plan_builder.go
+++ b/planner/core/logical_plan_builder.go
@@ -4471,30 +4471,33 @@ func (b *PlanBuilder) buildDataSource(ctx context.Context, tn *ast.TableName, as
 	}
 
 	if tableInfo.GetPartitionInfo() != nil {
-		h := domain.GetDomain(b.ctx).StatsHandle()
-		tblStats := h.GetTableStats(tableInfo)
-		isDynamicEnabled := b.ctx.GetSessionVars().IsDynamicPartitionPruneEnabled()
-		globalStatsReady := tblStats.IsInitialized()
-		// If dynamic partition prune isn't enabled or global stats is not ready, we won't enable dynamic prune mode in query
-		usePartitionProcessor := !isDynamicEnabled || !globalStatsReady
+		// If `UseDynamicPruneMode` already been false, then we don't need to check whether execute `flagPartitionProcessor`
+		// otherwise we need to check global stats initialized for each partition table
+		if b.ctx.GetSessionVars().StmtCtx.UseDynamicPruneMode {
+			h := domain.GetDomain(b.ctx).StatsHandle()
+			tblStats := h.GetTableStats(tableInfo)
+			isDynamicEnabled := b.ctx.GetSessionVars().IsDynamicPartitionPruneEnabled()
+			globalStatsReady := tblStats.IsInitialized()
+			// If dynamic partition prune isn't enabled or global stats is not ready, we won't enable dynamic prune mode in query
+			usePartitionProcessor := !isDynamicEnabled || !globalStatsReady
 
-		failpoint.Inject("forceDynamicPrune", func(val failpoint.Value) {
-			if val.(bool) {
+			failpoint.Inject("forceDynamicPrune", func(val failpoint.Value) {
+				if val.(bool) {
+					if isDynamicEnabled {
+						usePartitionProcessor = false
+					}
+				}
+			})
+
+			if usePartitionProcessor {
+				b.optFlag = b.optFlag | flagPartitionProcessor
+				b.ctx.GetSessionVars().StmtCtx.UseDynamicPruneMode = false
 				if isDynamicEnabled {
-					usePartitionProcessor = false
+					b.ctx.GetSessionVars().StmtCtx.AppendWarning(
+						fmt.Errorf("disable dynamic pruning due to %s has no global stats", tableInfo.Name.String()))
 				}
 			}
-		})
-
-		if usePartitionProcessor {
-			b.optFlag = b.optFlag | flagPartitionProcessor
-			b.ctx.GetSessionVars().StmtCtx.UseDynamicPruneMode = false
-			if isDynamicEnabled {
-				b.ctx.GetSessionVars().StmtCtx.AppendWarning(
-					fmt.Errorf("disable dynamic pruning due to %s has no global stats", tableInfo.Name.String()))
-			}
 		}
-
 		pt := tbl.(table.PartitionedTable)
 		// check partition by name.
 		if len(tn.PartitionNames) > 0 {

--- a/planner/core/logical_plan_builder.go
+++ b/planner/core/logical_plan_builder.go
@@ -4473,32 +4473,36 @@ func (b *PlanBuilder) buildDataSource(ctx context.Context, tn *ast.TableName, as
 	if tableInfo.GetPartitionInfo() != nil {
 		// If `UseDynamicPruneMode` already been false, then we don't need to check whether execute `flagPartitionProcessor`
 		// otherwise we need to check global stats initialized for each partition table
-		if b.ctx.GetSessionVars().StmtCtx.UseDynamicPruneMode {
-			h := domain.GetDomain(b.ctx).StatsHandle()
-			tblStats := h.GetTableStats(tableInfo)
-			isDynamicEnabled := b.ctx.GetSessionVars().IsDynamicPartitionPruneEnabled()
-			globalStatsReady := tblStats.IsInitialized()
-			// If dynamic partition prune isn't enabled or global stats is not ready, we won't enable dynamic prune mode in query
-			usePartitionProcessor := !isDynamicEnabled || !globalStatsReady
+		if !b.ctx.GetSessionVars().IsDynamicPartitionPruneEnabled() {
+			b.optFlag = b.optFlag | flagPartitionProcessor
+		} else {
+			if !b.ctx.GetSessionVars().StmtCtx.UseDynamicPruneMode {
+				b.optFlag = b.optFlag | flagPartitionProcessor
+			} else {
+				h := domain.GetDomain(b.ctx).StatsHandle()
+				tblStats := h.GetTableStats(tableInfo)
+				isDynamicEnabled := b.ctx.GetSessionVars().IsDynamicPartitionPruneEnabled()
+				globalStatsReady := tblStats.IsInitialized()
+				// If dynamic partition prune isn't enabled or global stats is not ready, we won't enable dynamic prune mode in query
+				usePartitionProcessor := !isDynamicEnabled || !globalStatsReady
 
-			failpoint.Inject("forceDynamicPrune", func(val failpoint.Value) {
-				if val.(bool) {
+				failpoint.Inject("forceDynamicPrune", func(val failpoint.Value) {
+					if val.(bool) {
+						if isDynamicEnabled {
+							usePartitionProcessor = false
+						}
+					}
+				})
+
+				if usePartitionProcessor {
+					b.optFlag = b.optFlag | flagPartitionProcessor
+					b.ctx.GetSessionVars().StmtCtx.UseDynamicPruneMode = false
 					if isDynamicEnabled {
-						usePartitionProcessor = false
+						b.ctx.GetSessionVars().StmtCtx.AppendWarning(
+							fmt.Errorf("disable dynamic pruning due to %s has no global stats", tableInfo.Name.String()))
 					}
 				}
-			})
-
-			if usePartitionProcessor {
-				b.optFlag = b.optFlag | flagPartitionProcessor
-				b.ctx.GetSessionVars().StmtCtx.UseDynamicPruneMode = false
-				if isDynamicEnabled {
-					b.ctx.GetSessionVars().StmtCtx.AppendWarning(
-						fmt.Errorf("disable dynamic pruning due to %s has no global stats", tableInfo.Name.String()))
-				}
 			}
-		} else {
-			b.optFlag = b.optFlag | flagPartitionProcessor
 		}
 		pt := tbl.(table.PartitionedTable)
 		// check partition by name.

--- a/statistics/handle/handle.go
+++ b/statistics/handle/handle.go
@@ -973,8 +973,13 @@ func (h *Handle) GetTableStats(tblInfo *model.TableInfo, opts ...TableStatsOpt) 
 
 // GetPartitionStats retrieves the partition stats from cache.
 func (h *Handle) GetPartitionStats(tblInfo *model.TableInfo, pid int64, opts ...TableStatsOpt) *statistics.Table {
-	statsCache := h.statsCache.Load().(statsCache)
 	var tbl *statistics.Table
+	if h == nil {
+		tbl = statistics.PseudoTable(tblInfo)
+		tbl.PhysicalID = pid
+		return tbl
+	}
+	statsCache := h.statsCache.Load().(statsCache)
 	var ok bool
 	option := &tableStatsOption{}
 	for _, opt := range opts {

--- a/statistics/handle/handle.go
+++ b/statistics/handle/handle.go
@@ -973,13 +973,8 @@ func (h *Handle) GetTableStats(tblInfo *model.TableInfo, opts ...TableStatsOpt) 
 
 // GetPartitionStats retrieves the partition stats from cache.
 func (h *Handle) GetPartitionStats(tblInfo *model.TableInfo, pid int64, opts ...TableStatsOpt) *statistics.Table {
-	var tbl *statistics.Table
-	if h == nil {
-		tbl = statistics.PseudoTable(tblInfo)
-		tbl.PhysicalID = pid
-		return tbl
-	}
 	statsCache := h.statsCache.Load().(statsCache)
+	var tbl *statistics.Table
 	var ok bool
 	option := &tableStatsOption{}
 	for _, opt := range opts {

